repo/blob/azure/azure_storage.go

```diff
@@ -0,0 +1,255 @@
+// Package azure implements Azure Blob Storage.
+package azure
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"github.com/efarrer/iothrottler"
+	"github.com/pkg/errors"
+	gblob "gocloud.dev/blob"
+	"gocloud.dev/blob/azureblob"
+	"gocloud.dev/gcerrors"
+
+	"github.com/kopia/kopia/internal/retry"
+	"github.com/kopia/kopia/repo/blob"
+)
+
+const (
+	azStorageType = "azureBlob"
+)
+
+type azStorage struct {
+	Options
+
+	ctx context.Context
+
+	bucket *gblob.Bucket
+
+	downloadThrottler *iothrottler.IOThrottlerPool
+	uploadThrottler   *iothrottler.IOThrottlerPool
+}
+
+func (az *azStorage) GetBlob(ctx context.Context, b blob.ID, offset, length int64) ([]byte, error) {
+	if offset < 0 {
+		return nil, errors.Errorf("invalid offset")
+	}
+
+	attempt := func() (interface{}, error) {
+		reader, err := az.bucket.NewRangeReader(ctx, az.getObjectNameString(b), offset, length, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		defer reader.Close() //nolint:errcheck
+
+		throttled, err := az.downloadThrottler.AddReader(reader)
+		if err != nil {
+			return nil, err
+		}
+
+		return ioutil.ReadAll(throttled)
+	}
+
+	v, err := exponentialBackoff(fmt.Sprintf("GetBlob(%q,%v,%v)", b, offset, length), attempt)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	fetched := v.([]byte)
+	if len(fetched) != int(length) && length >= 0 {
+		return nil, errors.Errorf("invalid offset/length")
+	}
+
+	return fetched, nil
+}
+
+func exponentialBackoff(desc string, att retry.AttemptFunc) (interface{}, error) {
+	return retry.WithExponentialBackoff(desc, att, isRetriableError)
+}
+
+func isRetriableError(err error) bool {
+	if me, ok := err.(azblob.ResponseError); ok {
+		if me.Response() == nil {
+			return true
+		}
+		// retry on server errors, not on client errors
+		return me.Response().StatusCode >= 500
+	}
+
```

Comment:
When would this return anything other than `false` beyond this point?

Reply:
Aah, default case should return `true` in the switch cases

Reply:
Let's add a test for this.

Reply:
Hi julio-lopez,
I have refactored the switch cases for checking error codes. We return `true` only if it is internal server error. I couldn't mock the scenario in the tests because we can only change creds. There is no way to override the default endpoint. I have also added few test negative cases.